### PR TITLE
Eliminate busy wait

### PR DESCRIPTION
**Pull request by arjo129**
Eliminates the busy wait previously used in `Transport` by replacing it with a custom `rclcpp::Executor` .

### CODE REVIEW

**Review by mxgrey — COMMENTED** [review]
The `Executor::wait_for_work` function is absolutely the key piece that was missing, so thanks for identifying that :+1: 

There are some issues with this implementation, though. The most significant issue is that we're running `wait_for_work()` in the same job that is being performed by the main `worker`. If I understand rxcpp correctly, that means the main `worker` will be blocked from doing anything else while waiting for new work to show up for the executor. That's a problem because the main `worker` is also used to synchronize the outputs of the planners and negotiators, so those outputs will be arbitrarily blocked while waiting for executor work.

I'm also unclear on the purpose of waiting for the back element of `weak_nodes_` as I mention in the inline comment.

I'll go ahead and create a PR to target this one with some changes that I think would improve the behaviors here.

File: rmf_fleet_adapter/rmf_rxcpp/include/rmf_rxcpp/Transport.hpp [hunk 1]
```diff
@@ -26,6 +26,61 @@
...
+    _worker.schedule(
+          [w = this](const auto&)
+    {
+      if (w->weak_nodes_.size() == 0)
+      {
+        w->_do_spin();
+        return;
+      }
+      //Assume at most 1 node. OK for our use case. 
+      else if (const auto lock = w->weak_nodes_.back().lock())
```
[the remark this answers is not in the document]
Comment (arjo129) [reply rmf_fleet_adapter/rmf_rxcpp/include/rmf_rxcpp/Transport.hpp:77]: OK we don't actually need this then.

**Comment by mxgrey** [comment]
@koonpeng would you be able to test this out on the web CI (without the `transport_sleep` parameter) to see if this also provides an acceptable solution to the busy waiting problem?

If these changes fix the problem then we could revert the changes done [here](https://github.com/osrf/rmf_core/pull/253).

**Comment by koonpeng** [comment]
It won't be straight forward as the CI uses a nightly build on master. We can revert the hack job changes as the CI is not using it actually.

**Comment by arjo129** [comment]
I've reverted the changes of #253 .

**Review by mxgrey — APPROVED** [review]

**Pull request merged**